Resteasy tries to instantiate abstract classes since Quarkus 3.3

### Describe the bug

I use an abstract base class for the Rest Application class in an common library, Quarkus 3.3 is suddenly trying to instantiate this class, Quarkus 3.2 is working as intended.

As a workaround, set `quarkus.resteasy.ignore-application-classes` to `true`

### Expected behavior

Resteasy should not try to instantiate abstract classes and simply ignore them.

### Actual behavior

Resteasy tries to instantiate an abstract class, failing with an exception at startup:

```
2024-06-24 00:38:50,770 ERROR [io.qua.run.boo.StartupActionImpl] (Quarkus Main Thread) Error running Quarkus: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:118)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:104)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.ExceptionInInitializerError
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized0(Native Method)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized(Unsafe.java:1160)
	at java.base/jdk.internal.reflect.MethodHandleAccessorFactory.ensureClassInitialized(MethodHandleAccessorFactory.java:300)
	at java.base/jdk.internal.reflect.MethodHandleAccessorFactory.newConstructorAccessor(MethodHandleAccessorFactory.java:103)
	at java.base/jdk.internal.reflect.ReflectionFactory.newConstructorAccessor(ReflectionFactory.java:200)
	at java.base/java.lang.reflect.Constructor.acquireConstructorAccessor(Constructor.java:549)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:70)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:44)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:124)
	at let.me.reproduce.Main.main(Main.java:10)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	... 3 more
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
	... 16 more
Caused by: java.lang.RuntimeException: RESTEASY003325: Failed to construct public let.me.reproduce.AbstractApplication()
	at org.jboss.resteasy.core.ConstructorInjectorImpl.constructOutsideRequest(ConstructorInjectorImpl.java:218)
	at org.jboss.resteasy.core.ConstructorInjectorImpl.construct(ConstructorInjectorImpl.java:188)
	at io.quarkus.resteasy.common.runtime.QuarkusConstructorInjector.construct(QuarkusConstructorInjector.java:37)
	at org.jboss.resteasy.core.providerfactory.Utils.createProviderInstance(Utils.java:97)
	at org.jboss.resteasy.core.providerfactory.ResteasyProviderFactoryImpl.createProviderInstance(ResteasyProviderFactoryImpl.java:1265)
	at org.jboss.resteasy.core.ResteasyDeploymentImpl.createApplication(ResteasyDeploymentImpl.java:385)
	at org.jboss.resteasy.core.ResteasyDeploymentImpl.initializeObjects(ResteasyDeploymentImpl.java:239)
	at org.jboss.resteasy.core.ResteasyDeploymentImpl.startInternal(ResteasyDeploymentImpl.java:136)
	at org.jboss.resteasy.core.ResteasyDeploymentImpl.start(ResteasyDeploymentImpl.java:124)
	at io.quarkus.resteasy.runtime.standalone.ResteasyStandaloneRecorder.staticInit(ResteasyStandaloneRecorder.java:57)
	at io.quarkus.deployment.steps.ResteasyStandaloneBuildStep$staticInit345281060.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.ResteasyStandaloneBuildStep$staticInit345281060.deploy(Unknown Source)
	... 17 more
Caused by: java.lang.InstantiationException
	at java.base/jdk.internal.reflect.InstantiationExceptionConstructorAccessorImpl.newInstance(InstantiationExceptionConstructorAccessorImpl.java:48)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
	at org.jboss.resteasy.core.ConstructorInjectorImpl.constructOutsideRequest(ConstructorInjectorImpl.java:203)
	... 28 more
```

### How to Reproduce?

1. Clone https://github.com/balta3/Quarkus-AbstractApplicationReproducer
2. Run the application, it will fail to start
3. To see this is a regression, set the Quarkus version to 3.2.12.Final and run the application, it will not fail to start

### Output of `uname -a` or `ver`

Linux tuxsgaming 6.9.5-1-default #1 SMP PREEMPT_DYNAMIC Tue Jun 18 07:38:24 UTC 2024 (c9c2e24) x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "21.0.3" 2024-04-16 OpenJDK Runtime Environment (build 21.0.3+9-suse-1.1-x8664) OpenJDK 64-Bit Server VM (build 21.0.3+9-suse-1.1-x8664, mixed mode, sharing)

### Quarkus version or git rev

3.3.3

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.8 Maven home: /usr/share/maven Java version: 21.0.3, vendor: N/A, runtime: /usr/lib64/jvm/java-21-openjdk-21 Default locale: de_DE, platform encoding: UTF-8 OS name: "linux", version: "6.9.5-1-default", arch: "amd64", family: "unix"

### Additional information

_No response_

--- a/extensions/resteasy-classic/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java
+++ b/extensions/resteasy-classic/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java
@@ -240,7 +240,9 @@ public class ResteasyServerCommonProcessor {
             jaxrsProvidersToRegisterBuildItem = getFilteredJaxrsProvidersToRegisterBuildItem(
                     jaxrsProvidersToRegisterBuildItem, allowedClasses, excludedClasses);
 
-            Collection<ClassInfo> knownApplications = index.getAllKnownSubclasses(ResteasyDotNames.APPLICATION);
+            Collection<ClassInfo> knownApplications = index.getAllKnownSubclasses(ResteasyDotNames.APPLICATION).stream()
+                    .filter(ci -> !ci.isAbstract()).collect(
+                            Collectors.toSet());
             // getAllowedClasses throws an Exception if multiple Applications are found, so we should only get 1
             if (knownApplications.size() == 1) {
                 appClass = knownApplications.iterator().next().name().toString();

--- a/extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/root/ApplicationTest.java
+++ b/extensions/resteasy-classic/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/root/ApplicationTest.java
@@ -42,7 +42,7 @@ class ApplicationTest {
                             ResponseFilter1.class, ResponseFilter2.class,
                             ResponseFilter3.class, ResponseFilter4.class, ResponseFilter5.class, ResponseFilter6.class,
                             Feature1.class, Feature2.class, DynamicFeature1.class, DynamicFeature2.class,
-                            ExceptionMapper1.class, ExceptionMapper2.class, AppTest.class));
+                            ExceptionMapper1.class, ExceptionMapper2.class, AppTest.class, AppTest2.class));
 
     @DisplayName("Should access to ok of resource 1 and provide a response with the expected headers")
     @Test
@@ -308,5 +308,9 @@ class ApplicationTest {
                     Arrays.asList(
                             new ResponseFilter1(), new DynamicFeature1()));
         }
+    }
+
+    public static abstract class AppTest2 extends Application {
+
     }
 }

